### ISSUE TITLE
change /predicates endpoint to /knowledge_map

### ISSUE THREAD

**Issue by vdancik** [NCATSTranslator/ReasonerAPI#143]
In addition to providing node types and predicates, we should also provide information about the types of node identifiers that the ARA/KP can consume and produce.

**Comment by edeutsch** [comment]
@vdancik there seems to be substantial interest in this. Do you want to put together a pull request with what you have in mind? I don't have a clear idea on what you're proposing. Perhaps something like the SRI NodeNormalizer's get_curie_prefixes endpoint?

https://nodenormalization-sri.renci.org/apidocs/#/Interfaces/get_get_curie_prefixes

**Comment by vdancik** [comment]
Here is an example knowledge map for ChEMBL target information. In addition to specifying predicates, node types, and node identifiers, information about attributes is also included.

```
{
    "knowledge_map": {
        "nodes": {
            "biolink:ChemicalSubstance": {
                "id_prefixes": [
                    "ChEMBL:",
                    "inchikey:"
                ],
                "attributes": [
                    {
                        "type": "data:0990",
                        "source": "ChEMBL",
                        "names": [
                            "name"
                        ]
                    },
                    {
                        "type": "data:1001",
                        "source": "ChEMBL",
                        "names": [
                            "synonym"
                        ]
                    }
                ]
            },
            "biolink:Disease": {
                "id_prefixes": [
                    "MESH:",
                    "EFO:"
                ],
                "attributes": [
                    {
                        "type": "data:3668",
                        "source": "ChEMBL",
                        "names": [
                            "name", "synonym"
                        ]
                    }
                ]
            }
        },
        "predicates": [
            {
                "subject": "biolink:ChemicalSubstance",
                "predicate": "biolink:treats",
                "target": "biolink:Disease",
                "source": "ChEMBL",
                "relations": [
                    "treats"
                ],
                "attributes": [
                    {
                        "type": "OPMI:0000367",
                        "source": "ChEMBL",
                        "names": [
                            "max phase for indication"
                        ]
                    },
                    {
                        "type": "NCIT:C25641",
                        "source": "ChEMBL",
                        "names": [
                            "ClinicalTrials","DailyMed","ATC","FDA"
                        ]
                    },
                ]
            }
        ]
    }
}
```

**Comment by southalln** [comment]
There might need to be a discussion of scoping whether this should be part of a resource's SmartAPI registration, included into the Reasoner API spec, or both. https://smart-api.info/api/metakg currently provides a similar service, derived from the "x-bte-kgs-operations", "x-bte-response-mapping", etc. annotations (for a simple example see: https://smart-api.info/api/metadata/1f277e1563fcfd124bfae2cc3c4bcdec ).

I like the extensions here linking back to the biolink semantic typing, and of providing additional attribute information for nodes/predicates.

What is still possibly missing is a sense for quantity or depth of content provided --- perhaps a simple count property would suffice as a first metric for resource richness.

**Comment by MarkDWilliams** [comment]
+1 to this idea as well as Noel's comments on it.

@newgene , do you have any thoughts on how much overlap there is between this and the expanded SmartAPI registration you've done for BTE?

**Comment by newgene** [comment]
I think we have two things to think about:

**1. What metadata fields are needed?**

   In terms of the necessary fields describing a KP's metakg, there are pretty much the same for reasoner APIs or other smartAPI KPs:

  * subject (semantic type, identifier type)
  * predicate (type, source, other attributes)
  * object (semantic type, identifier type)

   I agree we should synchronize as much as we can how these metadata are recorded. The above metadata are essential for the global metakg graph, so we should consider they are required.

  In addition to these fields, we will need some extra attributes related to how to make API call to get back the data for the requested predicate. I imagine Reasoner API and other SmartAPI KPs will have different sets of fields (e.g. handling parameter passing, batch support etc). We can try to synchronize on the naming convention, but it's less important.

**2. where to put these metadata**
   For reasoner API, this kind of metadata can be returned directly from the /predicates (or /knowledge_map) endpoint (as planned in this issue). SmartAPI registry will process the output and merge the metadata, with some transformation if needed, with the rest of metakg. Only thing SmartAPI registry requires for reasoner API is to include a "reasoner" tag in the API metadata.

  Since reasoner API is meant to be a standardized API, I would prefer to keeping these metadata within the reasoner specs, so that the developers don't have to update their API metadata when data changes. As long as SmartAPI registry knows it's a reasoner API, it's enough for the registry to process it properly.

**Comment by deepakunni3** [comment]
Here is a sample of KG metadata generated by KGX:
```
graph_name: Graph
node_stats:
  total_nodes: 14759537
  node_categories:
  - biolink:AnatomicalEntity
  - biolink:BiologicalProcess
  - biolink:CellularComponent
  - biolink:ChemicalSubstance
  - biolink:MolecularActivity
  - biolink:OntologyClass
  count_by_category:
    biolink:AnatomicalEntity:
      count: 67411
      provided_by:
        bgee:
          count: 3927
        coriell:
          count: 12
        mgi:
          count: 3
        mmrrc:
          count: 42614
        monarch:
          count: 14
        monarch-ontologies:
          count: 24798
    biolink:BiologicalProcess:
      count: 47548
      provided_by:
        flybase:
          count: 4
        go:
          count: 17416
        gwascatalog:
          count: 68
        kegg:
          count: 538
        monarch-ontologies:
          count: 30825
        reactome:
          count: 15910
        zfin:
          count: 278
    biolink:CellularComponent:
      count: 4464
      provided_by:
        go:
          count: 2325
        monarch-ontologies:
          count: 4464
    biolink:MolecularActivity:
      count: 12231
      provided_by:
        go:
          count: 5587
        monarch-ontologies:
          count: 12231
    biolink:OntologyClass:
      count: 149188
      provided_by:
        animalqtldb:
          count: 25
        bgee:
          count: 1033
        biogrid:
          count: 21
        chebi:
          count: 25
        clinvar:
          count: 16
        coriell:
          count: 30
        ctd:
          count: 11
edge_stats:
  total_edges: 35578700
  edge_labels:
  - biolink:actively_involved_in
  - biolink:affects
  - biolink:affects_localization_of
  - biolink:biomarker_for
  count_by_edge_label:
    biolink:actively_involved_in:
      count: 2261221
      provided_by:
        go:
          count: 1133614
        kegg:
          count: 81602
        reactome:
          count: 1046005
    biolink:affects:
      count: 24218
      provided_by:
        monarch-ontologies:
          count: 24218
    biolink:affects_localization_of:
      count: 6
      provided_by:
        monarch-ontologies:
          count: 6
    biolink:biomarker_for:
      count: 649384
      provided_by:
        animalqtldb:
          count: 584426
        ctd:
          count: 64341
        omim:
          count: 201
        orphanet:
          count: 416
  count_by_spo:
    biolink:AnatomicalEntity-biolink:affects-biolink:AnatomicalEntity:
      count: 8338
      provided_by:
        monarch-ontologies:
          count: 8338
    biolink:BiologicalProcess-biolink:affects_localization_of-biolink:CellularComponent:
      count: 6
      provided_by:
        monarch-ontologies:
          count: 6
    biolink:Cell-biolink:affects-biolink:Cell:
      count: 943
      provided_by:
        monarch-ontologies:
          count: 943
    biolink:OntologyClass-biolink:affects-biolink:BiologicalProcess:
      count: 4917
      provided_by:
        monarch-ontologies:
          count: 4917
    biolink:OntologyClass-biolink:affects-biolink:CellularComponent:
      count: 604
      provided_by:
        monarch-ontologies:
          count: 604
    biolink:OntologyClass-biolink:affects-biolink:MolecularActivity:
      count: 407
      provided_by:
        monarch-ontologies:
          count: 407
    biolink:OntologyClass-biolink:affects-biolink:OntologyClass:
      count: 9009
      provided_by:
        monarch-ontologies:
          count: 9009

```

**Comment by edeutsch** [comment]
Implemented at /meta_knowledge_graph in TRAPI 1.1
closing
